# Diagnósticos no espaço de embeddings

Vamos explorar embeddings de imagens, vetores de números que descrevem o conteúdo de imagens.
Com esses vetores dá para procurar cópias, encontrar imagens fora do padrão, agrupar o conjunto
e escolher o que vale a pena anotar.

In [1]:
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from PIL import Image, ImageOps

# Altere para o caminho do seu sistema
dataset_dir = Path("../dados/pvd_images")
images_dir = dataset_dir / "images"

In [ ]:
def load_rgb(path):
    """Abre qualquer imagem do conjunto e devolve um RGB de 8 bits. Essa função foi implementada
    no notebook anterior."""
    with Image.open(path) as image:
        image = ImageOps.exif_transpose(image)
        if image.mode in ("RGBA", "LA") or "transparency" in image.info:
            image = image.convert("RGBA")
            background = Image.new("RGBA", image.size, (255, 255, 255, 255))
            image = Image.alpha_composite(background, image)
        return image.convert("RGB")


def show_grid(names, titles=None, columns=6, size=1.9):
    """Mostra uma grade de imagens do conjunto, com um título opcional em cada uma."""

    # Calcula o número de linhas. ceil é usado caso a última linha tenha menos imagens.
    rows = int(np.ceil(len(names) / columns))
    fig, axes = plt.subplots(rows, columns, figsize=(columns*size, rows*size))
    axes = axes.flatten()

    for idx, name in enumerate(names):
        ax = axes[idx]
        title = titles[idx] if titles is not None else name
        ax.axis("off")
        ax.imshow(load_rgb(images_dir / name))
        ax.set_title(title, fontsize=8)

    fig.tight_layout(); plt.show()

## Carregar as features

O arquivo `features_dinov3.npz` guarda três coisas: o nome de cada imagem, a matriz de
vetores, e o nome do modelo que gerou os dados. 

In [3]:
data = np.load(dataset_dir / "features_dinov3.npz")

image_names = list(data["image_names"])
features = data["features"]

print(f"modelo ..... {str(data['model_name'])}")
print(f"features ... {features.shape}  ({features.dtype})")

modelo ..... vit_small_patch16_dinov3.lvd1689m
features ... (1845, 384)  (float32)


Cada linha é uma imagem e cada coluna é uma das 384 dimensões do espaço. 

<div class="alert alert-block alert-warning">
Notebook ainda não finalizado.
</div>